compiler/noirc_frontend/src/ast/type_alias.rs

```diff
@@ -4,18 +4,89 @@ use noirc_errors::Location;
 use std::fmt::Display;
 
 /// Ast node for type aliases
+/// A Noir Type Alias can be an alias to a normal type, or to a numeric generic type.
 #[derive(Clone, Debug)]
-pub struct NoirTypeAlias {
+pub enum NoirTypeAlias {
+    NormalTypeAlias(NormalTypeAlias),
+    NumericTypeAlias(NumericTypeAlias),
+}
+
+impl Display for NoirTypeAlias {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NoirTypeAlias::NormalTypeAlias(alias) => write!(f, "{}", alias),
+            NoirTypeAlias::NumericTypeAlias(alias) => write!(f, "{}", alias),
+        }
+    }
+}
+#[derive(Clone, Debug)]
+pub struct NormalTypeAlias {
     pub name: Ident,
     pub generics: UnresolvedGenerics,
     pub typ: UnresolvedType,
     pub visibility: ItemVisibility,
     pub location: Location,
 }
 
-impl Display for NoirTypeAlias {
+impl Display for NormalTypeAlias {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         let generics = vecmap(&self.generics, |generic| generic.to_string());
         write!(f, "type {}<{}> = {}", self.name, generics.join(", "), self.typ)
     }
 }
+
+#[derive(Clone, Debug)]
+pub struct NumericTypeAlias {
+    pub type_alias: NormalTypeAlias,
+    pub numeric_type: UnresolvedType,
```

Comment:
If the only difference between the two aliases is this numeric_type, maybe it could be turned into `numeric_type: Option<UnresolvedType>` in `NoirTypeAlias` to avoid an enum. I think parsing will be simplified too, and maybe the elaborator side too. The only downside I can see is that we need to remember to check this field during elaboration, but it's a one-time case and with a test we'll have it covered.

Reply:
I also think the `Option<UnresolvedType>` approach may be simpler

Reply:
Done